src/libsyntax/feature_gate.rs

```diff
@@ -2196,6 +2199,15 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
             gate_feature_post!(&self, c_variadic, span, "C-variadic functions are unstable");
         }
 
+        if fn_decl.inputs.iter().any(|x| !x.attrs.is_empty()) {
+            gate_feature_post!(
```

Comment:
Strictly speaking, feature gating for these attributes should be performed before macro expansion because they change what parser accepts (see https://github.com/rust-lang/rust/pull/48942/commits/7d5c29b9eae5857c040bf6f1b2d729596c8af3ae for an example of such feature gating).

This code, however, performs the check after macro expansion.
So the attributes will still be accepted on functions under `#[cfg(FALSE)]` without feature (so code on stable may break if we remove the feature in the future).

This feature, however, is unlikely to be removed, so it's kind of ok to leave the feature gating as is.
I'll leave this to @Centril to decide.

Reply:
@petrochenkov How difficult is it to feature gate it "properly"?

I don't mind doing it post expansion as done now but others in @rust-lang/lang  might have opinions.

Reply:
Not difficult, see the example in the linked commit https://github.com/rust-lang/rust/commit/7d5c29b9eae5857c040bf6f1b2d729596c8af3ae.

Reply:
In that case it seems justified to gate pre-expansion then.